File: gitdot-web/packages/api/src/endpoint/review/submit-review.ts
import { z } from "zod";
import { ReviewResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const SubmitReviewComment = z.object({
  body: z.string(),
  parent_id: z.uuid().nullable().optional(),
  file_path: z.string().nullable().optional(),
  line_number_start: z.number().int().nullable().optional(),
  line_number_end: z.number().int().nullable().optional(),
  side: z.string().nullable().optional(),
});
export type SubmitReviewComment = z.infer<typeof SubmitReviewComment>;

export const SubmitReviewRequest = z.object({
  action: z.string(),
  comments: z.array(SubmitReviewComment),
});
export type SubmitReviewRequest = z.infer<typeof SubmitReviewRequest>;

export const SubmitReviewResponse = ReviewResource;
export type SubmitReviewResponse = z.infer<typeof SubmitReviewResponse>;

export const SubmitReview = {
  path: "/repository/{owner}/{repo}/review/{number}/diff/{position}/submit",
  method: "POST",
  request: SubmitReviewRequest,
  response: SubmitReviewResponse,
} as const satisfies Endpoint;
export type SubmitReview = typeof SubmitReview;

File: gitdot-api/src/endpoint/review/submit_review.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{endpoint::Endpoint, resource::review::ReviewResource};

pub struct SubmitReview;

impl Endpoint for SubmitReview {
    const PATH: &'static str = "/repository/{owner}/{repo}/review/{number}/diff/{position}/submit";
    const METHOD: http::Method = http::Method::POST;

    type Request = SubmitReviewRequest;
    type Response = SubmitReviewResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct SubmitReviewRequest {
    pub action: String,
    pub comments: Vec<SubmitReviewComment>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SubmitReviewComment {
    pub body: String,
    pub parent_id: Option<Uuid>,
    pub file_path: Option<String>,
    pub line_number_start: Option<i32>,
    pub line_number_end: Option<i32>,
    pub side: Option<String>,
}

pub type SubmitReviewResponse = ReviewResource;
